File: app/src/main/java/org/example/example/fabric/PieceFabricator.java
package org.example.example.fabric;

import org.example.example.game.Cordinate;
import org.example.example.movement.InfiniteMove;
import org.example.example.movement.MovementBehaviour;
import org.example.example.movement.NormalMove;
import org.example.example.movement.SpecialMovement;
import org.example.example.piece.Color;
import org.example.example.piece.Piece;

import java.util.ArrayList;
import java.util.List;

public class PieceFabricator {

    public static Piece createPawn(Color color, Cordinate[] cordinate) {
        List<NormalMove> normalMoves = new ArrayList<>();
        List<SpecialMovement> specialMovements = new ArrayList<>();
        List<InfiniteMove> infiniteMoves = new ArrayList<>();
        if (color.equals(Color.WHITE)){
            normalMoves.add(new NormalMove(0,1, false));
            specialMovements.add(new SpecialMovement(0,2, false, cordinate));
        } else {
            normalMoves.add(new NormalMove(0,-1, false));
            specialMovements.add(new SpecialMovement(0,-2, false, cordinate));
        }
        return new Piece(color, new MovementBehaviour(normalMoves, specialMovements, infiniteMoves), "pawn", true);
    }

    public static Piece createKnight(Color color){
        List<NormalMove> normalMoves = new ArrayList<>();
        List<SpecialMovement> specialMovements = new ArrayList<>();
        List<InfiniteMove> infiniteMoves = new ArrayList<>();
        normalMoves.add(new NormalMove(1,2, true));
        normalMoves.add(new NormalMove(2,1, true));
        normalMoves.add(new NormalMove(2,-1, true));
        normalMoves.add(new NormalMove(1,-2, true));
        normalMoves.add(new NormalMove(-1,-2, true));
        normalMoves.add(new NormalMove(-2,-1, true));
        normalMoves.add(new NormalMove(-2,1, true));
        normalMoves.add(new NormalMove(-1,2, true));
        return new Piece(color, new MovementBehaviour(normalMoves, specialMovements, infiniteMoves), "knight", false);
    }

    public static Piece createRook(Color color){
        List<NormalMove> normalMoves = new ArrayList<>();
        List<SpecialMovement> specialMovements = new ArrayList<>();
        List<InfiniteMove> infiniteMoves = new ArrayList<>();
        infiniteMoves.add(new InfiniteMove(1,0));
        infiniteMoves.add(new InfiniteMove(-1,0));
        infiniteMoves.add(new InfiniteMove(0,1));
        infiniteMoves.add(new InfiniteMove(0,-1));
        return new Piece(color, new MovementBehaviour(normalMoves, specialMovements, infiniteMoves), "rook", false);
    }

    public static Piece createBishop(Color color){
        List<NormalMove> normalMoves = new ArrayList<>();
        List<SpecialMovement> specialMovements = new ArrayList<>();
        List<InfiniteMove> infiniteMoves = new ArrayList<>();
        infiniteMoves.add(new InfiniteMove(1,1));
        infiniteMoves.add(new InfiniteMove(-1,-1));
        infiniteMoves.add(new InfiniteMove(1,-1));
        infiniteMoves.add(new InfiniteMove(-1,1));
        return new Piece(color, new MovementBehaviour(normalMoves, specialMovements, infiniteMoves), "bishop", false);
    }

    public static Piece createQueen(Color color){
        List<NormalMove> normalMoves = new ArrayList<>();
        List<SpecialMovement> specialMovements = new ArrayList<>();
        List<InfiniteMove> infiniteMoves = new ArrayList<>();
        infiniteMoves.add(new InfiniteMove(1,0));
        infiniteMoves.add(new InfiniteMove(-1,0));
        infiniteMoves.add(new InfiniteMove(0,1));
        infiniteMoves.add(new InfiniteMove(0,-1));
        infiniteMoves.add(new InfiniteMove(1,1));
        infiniteMoves.add(new InfiniteMove(-1,-1));
        return new Piece(color, new MovementBehaviour(normalMoves, specialMovements, infiniteMoves), "queen", false);
    }

    public static Piece createKing(Color color, Cordinate[] cordinate){
        List<NormalMove> normalMoves = new ArrayList<>();
        List<SpecialMovement> specialMovements = new ArrayList<>();
        List<InfiniteMove> infiniteMoves = new ArrayList<>();
        normalMoves.add(new NormalMove(1,0, true));
        normalMoves.add(new NormalMove(-1,0, true));
        normalMoves.add(new NormalMove(0,1, true));
        normalMoves.add(new NormalMove(0,-1, true));
        normalMoves.add(new NormalMove(1,1, true));
        normalMoves.add(new NormalMove(-1,-1, true));
        specialMovements.add(new SpecialMovement(2,0, true, cordinate));
        specialMovements.add(new SpecialMovement(-2,0, true, cordinate));
        return new Piece(color, new MovementBehaviour(normalMoves, specialMovements, infiniteMoves), "king", false);
    }
}

File: app/src/main/java/org/example/fabric/PieceFabricator.kt
package org.example.fabric

import org.example.game.Validator.ValidatorMove
import org.example.game.Validator.move.DiagonalValidatorMove
import org.example.game.Validator.move.HorizontalValidatorMove
import org.example.game.Validator.move.JumpValidator
import org.example.game.Validator.move.VerticalValidatorMove
import org.example.game.Validator.pieces.JuggerValidatorPiece
import org.example.game.Validator.pieces.PawnValidatorMove
import org.example.game.Validator.special.CastleValidator
import org.example.game.piece.Color
import org.example.game.piece.Piece
import org.example.game.piece.PieceWithSpecialMove
import org.example.game.piece.PieceWithoutSpecialMove

class PieceFabricator {

    fun createPawn(id : Int , color : Color) : Piece{
        val hashMap = HashMap<String , Int>()

        hashMap.put("moves", 0)
        return PieceWithoutSpecialMove(id , color , "pawn" , listOf(PawnValidatorMove()) , hashMap, false, true)
    }

    fun createRook(id : Int , color : Color) : Piece{
        val hashMap = HashMap<String , Int>()

        hashMap.put("moves", 0)
        val validators = ArrayList<ValidatorMove>()

        validators.add(VerticalValidatorMove(0 , true))
        validators.add(VerticalValidatorMove(0 , false))
        validators.add(HorizontalValidatorMove(0 , true))
        validators.add(HorizontalValidatorMove(0 , false))
        return PieceWithoutSpecialMove(id , color , "rook" , validators , hashMap)
    }

    fun createKnight(id : Int , color : Color) : Piece{
        val hashMap = HashMap<String , Int>()
        val validators = ArrayList<ValidatorMove>()

        hashMap.put("moves", 0)
        validators.add(JumpValidator(1 , 2))
        validators.add(JumpValidator(2 , 1))
        validators.add(JumpValidator(1 , -2))
        validators.add(JumpValidator(2 , -1))
        validators.add(JumpValidator(-1 , 2))
        validators.add(JumpValidator(-2 , 1))
        validators.add(JumpValidator(-1 , -2))
        validators.add(JumpValidator(-2 , -1))

        return PieceWithoutSpecialMove(id , color , "knight" , validators , hashMap)
    }

    fun createBishop(id : Int , color : Color) : Piece{
        val hashMap = HashMap<String , Int>()

        hashMap.put("moves", 0)
        val validators = ArrayList<ValidatorMove>()
        validators.add(DiagonalValidatorMove(0 , listOf(1,2,3,4)))
        return PieceWithoutSpecialMove(id , color , "bishop" , validators , hashMap)
    }

    fun createQueen(id : Int , color : Color) : Piece{
        val hashMap = HashMap<String , Int>()

        hashMap.put("moves", 0)
        val validators = ArrayList<ValidatorMove>()
        validators.add(DiagonalValidatorMove(0 , listOf(1,2,3,4)))
        validators.add(VerticalValidatorMove(0 , true))
        validators.add(VerticalValidatorMove(0 , false))
        validators.add(HorizontalValidatorMove(0 , true))
        validators.add(HorizontalValidatorMove(0 , false))
        return PieceWithoutSpecialMove(id , color , "queen" , validators , hashMap)
    }

    fun createKing(id : Int , color : Color) : Piece{
        val hashMap = HashMap<String , Int>()

        hashMap.put("moves", 0)
        val validators = ArrayList<ValidatorMove>()
        validators.add(DiagonalValidatorMove(1 , listOf(1,2,3,4)))
        validators.add(VerticalValidatorMove(1 , true))
        validators.add(VerticalValidatorMove(1 , false))
        validators.add(HorizontalValidatorMove(1 , true))
        validators.add(HorizontalValidatorMove(1 , false))
        return PieceWithSpecialMove(id , color , "king" , validators , hashMap, listOf(CastleValidator()))
    }

    fun createKingWithouthCastle(id : Int , color: Color) : Piece{
        val hashMap = HashMap<String , Int>()

        hashMap.put("moves", 0)
        hashMap.put("kills", 0)
        val validators = ArrayList<ValidatorMove>()
        validators.add(DiagonalValidatorMove(1 , listOf(1,2,3,4)))
        validators.add(VerticalValidatorMove(1 , true))
        validators.add(VerticalValidatorMove(1 , false))
        validators.add(HorizontalValidatorMove(1 , true))
        validators.add(HorizontalValidatorMove(1 , false))
        return PieceWithoutSpecialMove(id , color , "king" , validators , hashMap)
    }

    fun createJugger(id : Int , color : Color) : Piece{
        val hashMap = HashMap<String , Int>()

        hashMap.put("moves", 0)
        val validators = ArrayList<ValidatorMove>()
        validators.add(JuggerValidatorPiece())
        return PieceWithoutSpecialMove(id , color , "jugger" , validators , hashMap)
    }

}